File: irrgarten/src/irrgarten/Weapon.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author marco
 */
public class Weapon {
    private float power;
    private int uses;
    
    // Constructor with parameters
    public Weapon(float power, int uses){
        this.power = power;
        this.uses = uses;
    }
    
    // Constructor without parameters, it set the attributes to 0
    public Weapon(){
        this(Dice.weaponPower(),Dice.usesLeft());
    }
    
    public float attack(){   
        
        float effectivePower = 0;
        
        if (uses > 0){
            effectivePower = power;
            uses--;
        }
        
        return effectivePower;
    }
    
    public String toString(){
        return "W[" + power + ", " + uses + "]";
    }
    
    public boolean discard(){
        return Dice.discardElement(uses);
    }
}

File: irrgarten_ruby/weapon.rb
# frozen_string_literal: true

module Irrgarten
  require_relative 'dice'
  class Weapon
    def initialize(power = Dice.weapon_power, uses = Dice.uses_left)
      @power = power
      @uses = uses
    end

    def attack
      effective_power = 0

      if @uses.positive?
        effective_power = @power
        @uses -= 1
      end

      effective_power
    end

    def to_s
      "W[#{@power}, #{@uses}]"
    end

    def discard
      Dice.discard_element(@uses)
    end
  end
end
